open-craft/opencraft
# [BB-1375] Re-add real-time updates to Ocim (with websockets through Django Channels)

Adds real-time updates to Ocim using Django channels.

**Testing instructions**:

The instructions given below are for testing in the vagrant environment and will require some changes to test in other environments.

* The `channels_redis` library used as the channel layer for `channels` requires Python 3.6 or above, but the version of Python available on Ubuntu 16.04 (used in production and on vagrant) only supports 3.5.x. So to test this PR, install `pyenv` and `pyenv-virtualenv` plugin. Install Python 3.6.9 (the latest Python 3.6 release), create a virtualenv using `pyenv-virtualenv`, activate it and install all the requirements in `requirements.txt.`,
* Setup the configuration in `.env` using the file in the private documentation repository. It is missing some configuration related to AWS - `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`. So add those variables with appropriate values.
* Run the unit test suite by running `make test.unit` from the top-level directory of this repository inside the vagrant VM to ensure that the environment is setup okay.
* Create a local superuser account if it doesn't exist already.
* Start the local Ocim instance by running `make run.dev`.
* Navigate to http://localhost:8000 and log in to the site using the super user credentials.
* Check the browser console to confirm that there are no errors.
* Create an instance using one of the factory methods.
* Verify that the instance automatically shows up in the Ocim interface.
* Click on the instance and launch an appserver. Immediately click on the `Logs` tab to check if the logs are updated in real time there.
* Also verify that there is a green circle near the instance name on the left side bar indicating that the instance is being provisioned. Later, check that the icon turns to red in real-time if the appserver fails to provision.
* Verify that after a few seconds, a new appserver shows up automatically under the instance.
* Click on the logs section of the appserver and verify that the logs are updated in real-time.
* Verify that there no JS errors in the browser console.
* Open the local Ocim website on a separate, isolated browser session where the user is unauthenticated. Open the browser JavaScript console and try creating a WebSocket client using the following code in the console.
  ```javascript
  var ws = new WebSocket('ws://localhost:8000/ws/');
  ```
  Verify that an error is thrown and the connection is rejected and that the websocket connections are allowed only for authenticated users.
* Check and test any other things of your choice related to real-time updates and verify that those work.

**Author notes and concerns**:
* All authenticated users are allowed to connect to the WebSocket endpoint `/ws/` and will receive all the websocket messages. The users with `manage_own` permission will also see the effects of the notifications for their instances alone. This mirrors the functionality in the removed swampdragon code. As a workaround, it might be better to allow only the superusers and the users with `manage_own` permissions to connect to the websocket endpoint. Any further changes to fix this issue will probably have to be done in a separate PR as a part of a separate ticket.
